{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Problem 1 solution\n",
    "germplasm = open(\"Germplasm.tsv\", \"r\")\n",
    "locusgene = open(\"LocusGene.tsv\", \"r\")\n",
    "\n",
    "a = 0\n",
    "counter = 0 #variable that counts the number of genes\n",
    "import re\n",
    "for line in germplasm.readlines():\n",
    "    germplasmAGIcode = re.search( r'AT\\dG\\d+\\t', line)\n",
    "    germplasm.seek(a)\n",
    "    if germplasmAGIcode:\n",
    "        for line in locusgene.readlines():\n",
    "            locusgeneAGIcode = re.search( r'AT\\dG\\d+\\t', line)\n",
    "            locusgene.seek(0)\n",
    "            a = a+1\n",
    "            if locusgeneAGIcode:\n",
    "                if germplasmAGIcode.group() == locusgeneAGIcode.group():\n",
    "                    print(locusgeneAGIcode.group()+\"is in both files\")\n",
    "                    counter+=1\n",
    "                \n",
    "germplasm.close()                    \n",
    "locusgene.close()                \n",
    "print (\"There are \"+str(counter)+\" genes\")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 1 explanation**\n",
    "\n",
    "For the problem number 1, I have used two loops for. The first one, read the file germplasm and looks for an AGIcode; after that, begin the second loop for that looks for AGI code in locusgene file and with an if command try to search the same AGIcode extracted from germplasm file.\n",
    "When the command if finish, the output prints the AGI code of the gene + \"is in both files\" to indicate that the gene is in germplasm and locusgene files.\n",
    "When the if command finishes, the code comeback to the first for loop, but begins the reading in the next line thanks to .seek() command. I used a variable called \"a\" that rise in 1 in each loop and, on this way, I get that the germplasm file reading begins a line after.\n",
    "At the end, with a variable called counter, I count the number of genes there are in common."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Problem 2 solution\n",
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "#%sql drop database Germplasm_Locusgene;\n",
    "            \n",
    "%sql create database Germplasm_Locusgene;           \n",
    "\n",
    "%sql use Germplasm_Locusgene;\n",
    "%sql CREATE TABLE Germplasm(Locus VARCHAR(9) NOT NULL, Germplasm VARCHAR(50) NOT NULL, Phenotype VARCHAR(1500) NOT NULL, Pubmed VARCHAR(50) NOT NULL);  \n",
    "%sql CREATE TABLE LocusGene(Locus VARCHAR(9) NOT NULL, Gene VARCHAR(20) NOT NULL, ProteinLength VARCHAR (20) NOT NULL)\n",
    "\n",
    "%sql show tables;\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 2 explanation**\n",
    "\n",
    "Using command in mysql(%sql) create database, I have created a database called Germplasm_Locusgene. After that, I have selected the database using a command called use and finally, I have created two tables inside the database Germplasm_Locusgene.\n",
    "The first table is called Germplasm and have four columns(Locus, Germplasm, Phenotype and Pubmed), the second table is called LocusGene and have three columns(Locus, Gene and ProteinLength).\n",
    "As both tables coincide in number and genes, the gene to gene relationship between tables is a 1:1 ratio, and I don't have to insert and auto primary key id."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Problem 3 solution\n",
    "germplasm = open(\"Germplasm.tsv\", \"r\")\n",
    "locusgene = open(\"LocusGene.tsv\", \"r\")\n",
    "locusgenelist = []\n",
    "germplasmlist = []\n",
    "\n",
    "for line in locusgene.readlines():\n",
    "    line = line.rstrip()\n",
    "    Locus, Gene, ProteinLength = line.split('\\t')\n",
    "    locusgenelist.append((Locus, Gene, ProteinLength)) #I do the list inserting data with .append() command\n",
    "for line in germplasm.readlines():\n",
    "    line = line.rstrip()\n",
    "    Locus, Germplasm, Phenotype, Pubmed = line.split('\\t')\n",
    "    germplasmlist.append((Locus, Germplasm, Phenotype, Pubmed))\n",
    "locusgene.close()\n",
    "germplasm.close()\n",
    "\n",
    "locusgenelist.pop(0)\n",
    "germplasmlist.pop(0)\n",
    "\n",
    "import pymysql.cursors\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Germplasm_Locusgene',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "#connection.autocommit = False\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql = \"\"\"INSERT INTO LocusGene (Locus, Gene, ProteinLength)\n",
    "            VALUES (%s,%s,%s)\"\"\"\n",
    "            val = locusgenelist  \n",
    "            cursor.executemany(sql,val) #execute sql and insert into %s the data contained in locusgenelist\n",
    "           \n",
    "            connection.commit()\n",
    "            \n",
    "            mql = \"\"\"INSERT INTO Germplasm (Locus, Germplasm, Phenotype, Pubmed)\n",
    "            VALUES (%s,%s,%s,%s)\"\"\"\n",
    "            cal = germplasmlist\n",
    "            cursor.executemany(mql, cal)\n",
    "            connection.commit()\n",
    "            \n",
    "finally:\n",
    "    print(\"\")\n",
    "connection.close()\n",
    "%sql SELECT * FROM Germplasm;\n",
    "%sql SELECT * FROM LocusGene;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM Germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 3 explanation**\n",
    "\n",
    "At first, I opened and read the tsv files using the command open(file, 'r'(read). After that I created two empty lists called locusgenelist and germplasmlist to fill them with two for loops. The for loop use the command readline to read the lines in order and fill the lists in order using the command .append. The file is a tsv file and for this reason I have used the command line.split(\\t) to shows the code how the elements in a line in the files are separated.\n",
    "Once I filled the lists, I close the files and remove the header line of each list using the command .pop(0) that remove the first element of the list (I have removed the header line).\n",
    "After that, using Python SQL library I filled the tables created in MySQL in problem 2 with the lists (locusgenelist and germplasmlist). To do it I filled the tables with command cursor.executemany and execute the sql line to insert data and the list to fill the values (%s).\n",
    "Finally, I have closed the connection and assure that the tables are filled correctly usig the command %sql SELECT * FROM.\n",
    "As we can see in the tables, the 32 genes are correctly inserted in the tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Problem 4.1 solution\n",
    "%sql SELECT * FROM LocusGene, Germplasm WHERE \\\n",
    "    LocusGene.Locus = Germplasm.Locus;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.1 explanation**\n",
    "\n",
    "To create a report that joined the Germplasm table and the LocusGene table, I have used the command SELECT * FROM LocusGene, Germplasm (to select both tables) WHERE LocusGene.Locus = Germplasm.Locus (to say to the code that the row Locus in LocusGene table is the same row that Locus row in Germplasm table).\n",
    "And on this way, I get an output that join both tables and ordering data according to the Locus.\n",
    "I didn't use id primary key because if I had used that, in the output would have had two id and two locus (one per table), but the tables have the same genes I can order tables in a relationship 1:1 using locus. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Problem 4.2 solution\n",
    "%sql SELECT * FROM LocusGene, Germplasm WHERE LocusGene.Locus = Germplasm.Locus AND (Gene = ('MAA3') OR Gene=('SKOR'));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.2 explanation**\n",
    "\n",
    "As I did in the problem 4.1, I used SELECT * FROM command and I put that LocusGene.Locus = Germplasm.Locus but I have shown only MAA3 and SKOR genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chromosome 1': 4}\n",
      "{'chromosome 2': 4}\n",
      "{'chromosome 3': 9}\n",
      "{'chromosome 4': 8}\n",
      "{'chromosome 5': 7}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Problem 4.3 solution\n",
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Germplasm_Locusgene',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "#connection.autocommit = False  # note that it is possible to delay putting changes into the database!\n",
    "#Report = open(\"Reports.tsv\", 'w')\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for chromosome in range(1,6):\n",
    "            chr_name= \"chromosome \"+str(chromosome)\n",
    "            regex_var = \"AT\"+str(chromosome)+\"G.+\"\n",
    "            sql= \"SELECT COUNT(*) AS %s FROM LocusGene, Germplasm WHERE \\\n",
    "                  LocusGene.Locus = Germplasm.Locus AND \\\n",
    "                 LocusGene.Locus REGEXP %s;\" \n",
    "            val= (chr_name, regex_var) #variable var contains data to insert in sql \n",
    "            cursor.execute(sql, val) #execute sql and insert val data into %s.\n",
    "            print(cursor.fetchone())\n",
    "            #report.write(str(cursor.fetchone()))\n",
    "           \n",
    "           \n",
    "                         \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "#report.close()\n",
    "\n",
    "#checkcontent = open(\"Reports.tsv\", \"r\")\n",
    "#print(checkcontent.read())\n",
    "#checkcontent.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.3 explanation**\n",
    "\n",
    "At first, I connected with the SQL library in python and use the database Germplasm_Locusgene.\n",
    "With a for loop in range (1,6), I selected the chromosomos from 1 to 5. In the loop, I have created two variables; the first one to name the chromosome and the second one to write the regular expression of the AGIcode per each chromosome.\n",
    "After that and using the command count in SQL I counted the number of genes there are in each chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Problem 4.4 solution\n",
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Germplasm_Locusgene',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "#connection.autocommit = False  # note that it is possible to delay putting changes into the database!\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        # Read a single record\n",
    "        for chromosome in range(1,6):\n",
    "            chr_name= \"chromosome \"+str(chromosome)\n",
    "            regex_var = \"AT\"+str(chromosome)+\"G.+\"\n",
    "            sql= \"SELECT AVG(LocusGene.ProteinLength) FROM LocusGene, Germplasm WHERE \\\n",
    "             LocusGene.Locus = Germplasm.Locus AND \\\n",
    "             LocusGene.Locus REGEXP %s;\"\n",
    "            val=(regex_var)\n",
    "            cursor.execute(sql, val)\n",
    "            print(cursor.fetchone())\n",
    "            \n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.4 explanation**\n",
    "Like in problem number 4.3 I used the library for SQL in python and the same loop for each chromosome. The difference between this one and the previous one is that I have used the command average (AVG).\n",
    "\n",
    "I couldn't have wroten the file using command Report = open(\"Reports.tsv\", 'w')\n",
    "report.write(str(cursor.fetchone()))\n",
    "report.close() (This commands are in problem 4.3)\n",
    "I thought that this way could be a great way to write the file reports but I don't know why the data didn't save.\n",
    "Instead of this way, I used the wrong way to make the file reports. I ran the code, copied the output and pasting the output into a tsv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
